Fix whole editor paste with locked code

KaiSaba opened the pull request:
…corations

CGNonofr commented on src/tools.ts:
```diff
@@ -28,6 +28,21 @@ function getRangesFromDecorations (
+  return firstUneditableRange != null
+    ? new monaco.Range(fullModelRange.startLineNumber, fullModelRange.startColumn, firstUneditableRange.startLineNumber, firstUneditableRange.startColumn)
+    : undefined
+}
```
So we only support a single range now? what is the point of computing multiple one then?

We only have some code to invert the ranges in this file around line 320

KaiSaba replied:
I added a generic function that recomputes the editor operations: f538ee0.

CGNonofr replied:
Can the code in the paste callback be removed then?

KaiSaba replied:
Yes, we should be able to remove the modification of `editor.trigger`. I'll removing it and test it.

KaiSaba replied:
Here is the commit that cleans up the code: 04bfcfb.

KaiSaba replied:
I found a bug while testing, here is the fix: c40f3c1.

CGNonofr replied:
Can you please split the code a little bit more? I'm struggling to understand the logic

KaiSaba replied:
Here is the refactor:  b598071.

I don't think I can split more the functions.

KaiSaba replied:
I moved the functions in other files: 8faaeb6.

I changed the logic to split the ranges, then split the operation text separately: c291e16.

KaiSaba replied:
Here are the commits that fix the last bugs: 8412173 and e47df09.

CGNonofr commented on src/tools.ts:
```diff
@@ -28,6 +17,123 @@ function getRangesFromDecorations (
+  let lastEndColumn = uniqueRange.startColumn
+
+  for (const range of ranges) {
+    const newRange = new monaco.Range(lastEndLineNumber, lastEndColumn, range.startLineNumber, range.startColumn)
```
nit: you can use a monaco.Position instead of the lineNumber and column

KaiSaba replied:
7ee4792

CGNonofr commented on src/tools.ts:
```diff
@@ -28,6 +17,123 @@ function getRangesFromDecorations (
+    // The operation range is entirely in an uneditable range
+    if (uneditableRangesThatIntersects.some(range => range.containsRange(operationRange))) {
+      displayLockedCodeError(
+        new monaco.Position(operationRange.startLineNumber, operationRange.startColumn))
```
It's probably not the responsability of this fonction to display an error

KaiSaba replied:
I removed the call to `displayLockedCodeError` while refactoring the function: b598071.

KaiSaba commented:
@CGNonofr I added 3 commits to refactor and fix the remaining bugs:
- f8cea32: refactor: refactor minusRanges function to return all ranges
- f8c1ed4: fix: use minusRanges to get hidden areas and visible ranges in hideCodeWithoutDecoration
- 0c5ae90: fix: fix split operation text function when uneditable range text not in text to split

KaiSaba commented:
@CGNonofr I added a commit to fix the line added on paste:
- 8c23db2: fix: fix locked code ranges and text split

CGNonofr commented on src/tests/utils.ts:
```diff
@@ -0,0 +1,58 @@
+  endColumn?: number
+): monaco.Range {
+  if (endColumn == null) {
+    endColumn = model.getLineMaxColumn(endLine)
```
nit: you can use it as default value line 18 (:

KaiSaba replied:
[test: simplify createTestRange function](https://github.com/CodinGame/monaco-editor-wrapper/pull/125/commits/8ec243327bd2e8a240bd3ab1e01f1207a3c3e81d)

CGNonofr commented on src/tools.ts:
```diff
@@ -201,6 +141,7 @@ function lockCodeUsingDecoration (
         return editorOperations
       }
 
+      editorOperations = computeNewOperationsForLockedCode(model, decorationFilter, editorOperations, withDecoration)
```
maybe we can find a more meaningful name :thinking: like `tryIgnoreLockedCode` or something?

KaiSaba replied:
[refactor: rename computeNewOperationsForLockedCode to tryIgnoreLockedCode](https://github.com/CodinGame/monaco-editor-wrapper/pull/125/commits/0759eb0e9e1235de24c289da9b02cc7e9a14580d)

CGNonofr commented on src/tools/utils/editorOperationUtils.ts:
```diff
@@ -0,0 +1,114 @@
+  model: monaco.editor.ITextModel,
+  uneditableRanges: monaco.Range[],
+  text: string | null
+): (string | null)[] {
```
how can it be null?

CGNonofr commented on src/tools/utils/editorOperationUtils.ts:
```diff
@@ -0,0 +1,114 @@
+  const uneditableRangesText = uneditableRanges.map(range => model.getValueInRange(range))
+  let currentRange: number = 0
+  let textToSplit: string = text
+  while (textToSplit !== '' && currentRange < uneditableRangesText.length) {
```
nit: `textToSplit.length > 0` seems clearer to me (we're looping until the string is empty)

also, wouldn't something like `remainingText` be more relevant?

KaiSaba replied:
[refactor: rename functions on locked code and their variables](https://github.com/CodinGame/monaco-editor-wrapper/pull/125/commits/4211c36f490bd0ebf871eeb479fac4afd597cb2d)

CGNonofr commented on src/tools/utils/editorOperationUtils.ts:
```diff
@@ -0,0 +1,114 @@
+
+function splitOperationText (
+  model: monaco.editor.ITextModel,
+  uneditableRanges: monaco.Range[],
```
`uneditableRangesInOperationRange`?

CGNonofr commented on src/tools/utils/editorOperationUtils.ts:
```diff
@@ -0,0 +1,114 @@
+    return editorOperations
+  }
+
+  return splitOperationsForLockedCode(model, editorOperations, uneditableRanges)
```
this function name should probably be similar to the called function name

KaiSaba replied:
[refactor: rename functions on locked code and their variables](https://github.com/CodinGame/monaco-editor-wrapper/pull/125/commits/4211c36f490bd0ebf871eeb479fac4afd597cb2d)

CGNonofr commented on src/tools/utils/editorOperationUtils.ts:
```diff
@@ -0,0 +1,114 @@
+  return splitText.map((text, index) => createNewOperation(oldOperation, editableRanges[index]!, text, index))
+}
+
+function splitOperationText (
```
What about `tryIgnoreLockedCodeForOperation`?

KaiSaba replied:
[refactor: rename functions on locked code and their variables](https://github.com/CodinGame/monaco-editor-wrapper/pull/125/commits/4211c36f490bd0ebf871eeb479fac4afd597cb2d)

CGNonofr commented on src/tools/utils/editorOperationUtils.ts:
```diff
@@ -0,0 +1,114 @@
+      const rangeTextIndex = textToSplit.indexOf(rangeText)
+
+      if (rangeTextIndex === -1) {
+        return []
```
What about throwing an error instead?

KaiSaba replied:
[fix: throw custom error when operation is invalid with locked code](https://github.com/CodinGame/monaco-editor-wrapper/pull/125/commits/bbf980c6092682d7f7c3cd531813a815c518a63c)

CGNonofr commented on src/tools/utils/editorOperationUtils.ts:
```diff
@@ -0,0 +1,114 @@
+  text: string | null
+): (string | null)[] {
+  if (text == null || text === '') {
+    return uneditableRanges.length > 0 ? [] : [text]
```
What about throwing an error instead here too?

KaiSaba replied:
[fix: throw custom error when operation is invalid with locked code](https://github.com/CodinGame/monaco-editor-wrapper/pull/125/commits/bbf980c6092682d7f7c3cd531813a815c518a63c)

CGNonofr commented on tsconfig.json:
```diff
@@ -19,7 +19,15 @@
+    "rollup.config.ts",
+    "jest.config.ts",
+    "src/tests",
+    "jest",
```
what is "jest"? also It seems src/tests is included in src/ :)

CGNonofr commented on src/tests/utils.ts:
```diff
@@ -0,0 +1,58 @@
+  ]
+}
+
+export function canTestOperationsEditRanges (splitOperations: ValidAnnotatedEditOperation[], uneditableRanges: monaco.Range[]): boolean {
```
I'm confused about the meaning of this function name :thinking:

CGNonofr commented on src/tests/editorOperationUtils/lockedCode.test.ts:
```diff
@@ -0,0 +1,192 @@
+
+    expect(splitOperations.length).toEqual(1)
+    expect(splitOperations[0]).toEqual(createTestOperation(operationRange, '// tata'))
+    expect(canTestOperationsEditRanges(splitOperations, uneditableRanges)).toBe(false)
```
what about applying the edit to the editor and check the final editor content?

KaiSaba commented on src/tools.ts:
```diff
@@ -28,6 +17,166 @@ function getRangesFromDecorations (
+  const startIndex = text.indexOf(rangeValue)
+  return {
+    startIndex,
+    endIndex: startIndex + rangeValue.length
```
This code was in an outdated commit and has been removed with the refactor.

KaiSaba commented:
@CGNonofr I added 2 commits:
- [cleanup: remove unneeded console.info](https://github.com/CodinGame/monaco-editor-wrapper/pull/125/commits/7c5a6752f3acff13885667ac95eab65e51807595)
- [fix: removed line bug when pasting code](https://github.com/CodinGame/monaco-editor-wrapper/pull/125/commits/b74967419885ed7391c1d5ab39e094f762d81b22)

CGNonofr commented on src/tools/utils/editorOperationUtils.ts:
```diff
@@ -2,6 +2,14 @@ import * as monaco from 'monaco-editor'
+  constructor (message: string) {
+    super(message)
+    this.name = 'LockedCodeError'
+    Object.setPrototypeOf(this, LockedCodeError.prototype)
```
What is it for?

CGNonofr commented on src/tools/utils/editorOperationUtils.ts:
```diff
@@ -26,7 +34,7 @@ function createNewOperationsFromRanges (
   editableRanges: monaco.Range[],
   splitText: (string | null)[]
 ): ValidAnnotatedEditOperation[] {
-  if (editableRanges.length <= 0 || splitText.length <= 0) {
```
why removing it?
